gnmi_server/clientCertAuth.go

```diff
@@ -44,9 +127,92 @@ func ClientCertAuthenAndAuthor(ctx context.Context, serviceConfigTableName strin
 		}
 	}
 
+	if enableCrl {
+		err := VerifyCertCrl(tlsAuth.State)
+		if err != nil {
+			glog.Infof("[%s] Failed to verify cert with CRL; %v", rc.ID, err)
+			return ctx, err
+		}
+	}
+
 	return ctx, nil
 }
 
+func TryDownload(url string) bool {
+	resp, err := http.Get(url)
+	defer resp.Body.Close()
+	if err != nil || resp.StatusCode != http.StatusOK {
+		glog.Infof("Download CRL: %s failed: %v", url, err)
+		return false
+	}
+
+	crlContent, err := io.ReadAll(resp.Body) 
+	if err != nil {
+		glog.Infof("Download CRL: %s failed: %v", url, err)
+		return false
+	}
+
+	AppendCrlToCache(url, crlContent)
+
+	return true
+}
+
+func GetCrlUrls(cert x509.Certificate) []string {
+	glog.Infof("Get Crl Urls for cert: %v", cert)
+	return cert.CRLDistributionPoints
+}
+
+func DownloadNotCachedCrl(crlUrlArray []string) bool {
+    for _, crlUrl := range crlUrlArray{
+		exist, _ := SearchCrlCache(crlUrl)
+		if !exist {
+			downloaded := TryDownload(crlUrl)
+			if !downloaded {
+				return false
+			}
+		}
+    }
+
+	return true
+}
+
+func CreateStaticCRLProvider() *StaticCRLProvider {
+	crlArray := make([][]byte, 1)
+	for mapkey, item := range(CrlCache) {
+		if CrlExpired(item) {
+			delete(CrlCache, mapkey)
+		} else {
+			crlArray = append(crlArray, item.crl)
+		}
+	}
+	
+	return NewStaticCRLProvider(crlArray)
+}
+
+func VerifyCertCrl(tlsConnState tls.ConnectionState) error {
+	// Check if any CRL already exist in local
+	crlUriArray := GetCrlUrls(*tlsConnState.VerifiedChains[0][0])
+	downloaded := DownloadNotCachedCrl(crlUriArray)
```

Comment:
If crl is enabled, but we failed to download crl, all the gnmi requests will be rejected.
Is this acceptable?

Reply:
It's by design:
If a cert does not contains any CRL distribution points, cert will not verify by CRL.
If a cert contains multiple CRL distribution points, and download from all distribution points failed, then the cert should be blocked, because we can't verify it.